src/Evolution/Systems/Cce/WorldtubeBufferUpdater.cpp

```diff
@@ -454,6 +454,94 @@ void BondiWorldtubeH5BufferUpdater::pup(PUP::er& p) {
   }
 }
 
+KleinGordonWorldtubeH5BufferUpdater::KleinGordonWorldtubeH5BufferUpdater(
+    const std::string& cce_data_filename,
+    const std::optional<double> extraction_radius)
+    : cce_data_file_{cce_data_filename}, filename_{cce_data_filename} {
+  get<Tags::detail::InputDataSet<
+      Spectral::Swsh::Tags::SwshTransform<Tags::KleinGordonPsi>>>(
+      dataset_names_) = "KGPsi";
+  get<Tags::detail::InputDataSet<
+      Spectral::Swsh::Tags::SwshTransform<Tags::KleinGordonPi>>>(
+      dataset_names_) = "dtKGPsi";
+
+  const std::string text_radius = detail::get_text_radius(cce_data_filename);
+  try {
+    extraction_radius_ = static_cast<bool>(extraction_radius)
+                             ? *extraction_radius
+                             : std::stod(text_radius);
+  } catch (const std::invalid_argument&) {
+    // the extraction radius is typically not used in the Klein-Gordon system,
+    // so we don't error if it isn't parsed from the filename. Instead, we'll
+    // just error if the invalid extraction radius value is ever retrieved using
+    // `get_extraction_radius`.
+  }
+
+  detail::set_time_buffer_and_lmax(make_not_null(&time_buffer_), l_max_,
+                                   cce_data_file_.get<h5::Dat>("/KGPsi"));
+  cce_data_file_.close_current_object();
+}
+
+// Same as the Bondi updater except for the tags (klein_gordon_input_tags)
+double KleinGordonWorldtubeH5BufferUpdater::update_buffers_for_time(
```

Comment:
If it's the same (except for tags), is there a way to factor it out (even with a template parameter)? We want to try and reduce code duplication to a minimum.